homeassistant/components/google_calendar.py

```diff
@@ -0,0 +1,407 @@
+
+_LOGGER = logging.getLogger(__name__)
+_CONFIGURING = {}
+
+DOMAIN = 'google_calendar'
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+
+CONF_FQDN = 'fqdn'
+CONF_TRACK_NEW = 'track_new_calendar'
+CONF_SCAN_INTERVAL = 'scan_interval'
+CONF_CAL_ID = 'cal_id'
+CONF_NAME = 'name'
+CONF_ENTITIES = 'entities'
+CONF_TRACK = 'track'
+CONF_SEARCH = 'search'
+CONF_OFFSET = 'offset'
+
+DEFAULT_CONF_TRACK_NEW = True
+DEFAULT_CONF_SCAN_INTERVAL = 60
+DEFAULT_CONF_OFFSET = '#-'
+
+DEFAULT_GOOGLE_SEARCH_PARAMS = {
+    'orderBy': 'startTime',
+    'maxResults': 1,
+    'singleEvents': True,
+}
+
+GROUP_NAME_ALL_CALENDARS = "Google Calendar Sensors"
+
+DISCOVER_BINARY_SENSORS = '{}.binary_sensor'.format(DOMAIN)
+SERVICE_SCAN_CALENDARS = 'scan_for_calendars'
+
+YAML_DEVICES = '{}_devices.yaml'.format(DOMAIN)
+SCOPES = 'https://www.googleapis.com/auth/calendar.readonly'
+CONFIG_FILE = '{}.conf'.format(DOMAIN)
+TOKEN_FILE = '.{}.token'.format(DOMAIN)
+CAL_SEARCH_START = '/auth/{}'.format(DOMAIN)
+CAL_SEARCH_CALLBACK = '/auth/{}/callback'.format(DOMAIN)
+
+# Return cached results if last scan was less then this time ago
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=5)
+
+PLATFORM_SCHEMA = vol.Schema({
+    vol.Optional(CONF_FQDN): cv.string,
+    vol.Optional(CONF_TRACK_NEW): cv.boolean,
+    vol.Optional(CONF_SCAN_INTERVAL): cv.positive_int,
+})
+
+_SINGLE_CALSEARCH_CONFIG = vol.Schema({
+    vol.Required(CONF_NAME): cv.string,
+    vol.Optional(CONF_TRACK): cv.boolean,
+    vol.Optional(CONF_SEARCH): vol.Any(cv.string, None),
+    vol.Optional(CONF_OFFSET): cv.string,
+})
+
+_CALENDARS = {}
+_OAUTH_PATH_SETUP = False
+
+
+def get_fqdn():
+    """Get system fqdn or None."""
+    import socket
+    from homeassistant.util import get_local_ip
+    hostname = 'PLEASE_SET_FQDN_IN_CONFIGURATION'
+    try:
+        hostname = socket.gethostbyaddr(get_local_ip())[0]
+    except socket.gaierror:
+        _LOGGER.error('Unable to get the hostname automatically, ' +
+                      'you need to set %s in the configuration',
+                      CONF_FQDN)
+    return hostname
+
+
+def _calsearch_dict(value):
+    """Validate a dictionary of group definitions."""
+    config = []
+    for calsearch in value:
+        try:
+            config.append(_SINGLE_CALSEARCH_CONFIG(calsearch))
+        except vol.MultipleInvalid as ex:
+            raise vol.Invalid('Google Calendar Entity is invalid: {}: {}'.
+                              format(calsearch, ex))
+    return config
+
+
+DEVICE_SCHEMA = vol.Schema({
+    vol.Required(CONF_CAL_ID): cv.string,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_ENTITIES, None): vol.All(list, _calsearch_dict),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup_oauth_paths(hass, config):
+    """Start setup of app with Google Oauth."""
+    # pylint disable=import-error,global-statement
+    base_url = "{}://{}:{}".format(
+        'https' if hass.http.use_ssl else 'http',
+        config.get(CONF_FQDN, get_fqdn()),
```

Comment:
Why not just use the local ip ? 

Also, if we want to add an `get_fqdn()` function, it should be added to `util/__init__.py`